File: web/src/main/kotlin/app/PresenterProvider.kt
package app

import de.compeople.swn.presentation.displayResults.DisplayResultsPresenter
import de.compeople.swn.presentation.displayResults.DisplayResultsView
import de.compeople.swn.presentation.userinput.UserInputPresenter
import de.compeople.swn.presentation.userinput.UserInputView
import de.compeople.swn.tarifService.Rechenkern
import de.compeople.swn.tarifService.TarifClient
import de.compeople.swn.tarifService.TarifRepository
import de.compeople.swn.tarifService.TarifService
import de.compeople.swn.time.TimeService

class PresenterProvider {

    private val timeService = TimeService()
    private val keyValueStore = KeyValueStoreWeb()
    private val tarifRepo = TarifRepository(keyValueStore)
    private val tarifService = TarifService(tarifRepo, TarifClient(), timeService)
    private val rechenkern = Rechenkern(tarifService, timeService)

    fun createUserInputPresenter(view: UserInputView): UserInputPresenter {
        return UserInputPresenter(view, rechenkern)
    }

    fun createDisplayResultsPresenter(view: DisplayResultsView): DisplayResultsPresenter {
        return DisplayResultsPresenter(view)
    }
}

File: native/iosApp/iosApp/PresenterProvider.swift
//
//  PresenterProvider.swift
//  iosApp
//

import Foundation
import commonClient

class PresenterProvider {

    private let keyValueStore:KeyValueStore
    private let timeService:CommonTimeService
    private let tarifRepo:TarifRepository
    private let tarifService:TarifService
    private let rechenkern: CommonRechenkern
    
    private init() {
        timeService = CommonTimeService()
        keyValueStore = KeyValueStoreIOS()
        tarifRepo = TarifRepository(keyValueStore: keyValueStore)
        tarifService = TarifService(tarifRepository: tarifRepo, tarifClient: TarifClient(),timeService: timeService)
        rechenkern = CommonRechenkern(tarifService: tarifService, timeService: timeService)
    }
    
    private static let instance = PresenterProvider()
    
    static func userInputPresenter(view:UserInputView) -> UserInputPresenter {
        return UserInputPresenter(view:view, rechenkern:instance.rechenkern)
    }
    
    static func displayResultsPresenter(view:DisplayResultsView) -> DisplayResultsPresenter {
        return DisplayResultsPresenter(displayResultsView: view)
    }
    
}
